lms/djangoapps/grades/tests/test_new.py

```diff
@@ -0,0 +1,223 @@
+        """
+        Tests to ensure that a persistent subsection grade is created, saved, then fetched on re-request.
+        """
+        with patch(
+            'lms.djangoapps.grades.new.subsection_grade.SubsectionGradeFactory._update_saved_grade',
+            wraps=self.subsection_grade_factory._update_saved_grade  # pylint: disable=protected-access
+        ) as mock_save_grades:
+            with patch(
+                'lms.djangoapps.grades.new.subsection_grade.SubsectionGradeFactory._get_saved_grade',
+                wraps=self.subsection_grade_factory._get_saved_grade  # pylint: disable=protected-access
+            ) as mock_get_saved_grade:
+                with self.assertNumQueries(17):
+                    grade_a = self.subsection_grade_factory.create(self.sequence, self.course_structure, self.course)
+                self.assertTrue(mock_get_saved_grade.called)
+                self.assertTrue(mock_save_grades.called)
+
+                mock_get_saved_grade.reset_mock()
+                mock_save_grades.reset_mock()
+
+                with self.assertNumQueries(3):
+                    grade_b = self.subsection_grade_factory.create(self.sequence, self.course_structure, self.course)
+                self.assertTrue(mock_get_saved_grade.called)
+                self.assertFalse(mock_save_grades.called)
+
+        self.assertEqual(grade_a.url_name, grade_b.url_name)
+        self.assertEqual(grade_a.all_total, grade_b.all_total)
+
+    @ddt.data(
+        (True, True),
+        (True, False),
+        (False, True),
+        (False, False),
+    )
+    @ddt.unpack
+    def test_subsection_grade_feature_gating(self, feature_flag, course_setting):
+        # Grades are only saved if the feature flag and the advanced setting are
+        # both set to True.
+        with patch(
+            'lms.djangoapps.grades.models.PersistentSubsectionGrade.read_grade'
+        ) as mock_read_saved_grade:
+            with patch.dict(settings.FEATURES, {'ENABLE_SUBSECTION_GRADES_SAVED': feature_flag}):
+                with patch.object(self.course, 'enable_subsection_grades_saved', new=course_setting):
+                    self.subsection_grade_factory.create(self.sequence, self.course_structure, self.course)
+        self.assertEqual(mock_read_saved_grade.called, feature_flag and course_setting)
+
+
+class SubsectionGradeTest(GradeTestBase):
+    """
+    Tests SubsectionGrade functionality.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super(SubsectionGradeTest, cls).setUpClass()
+
+    def setUp(self):
+        super(SubsectionGradeTest, self).setUp()
+
+    def test_compute(self):
+        """
+        Assuming the underlying score reporting methods work, test that the score is calculated properly.
+        """
+        grade = self.subsection_grade_factory.create(self.sequence, self.course_structure, self.course)
+        with patch('lms.djangoapps.grades.new.subsection_grade.get_score', return_value=(0, 1)):
```

Comment:
It would be great to add this as another generally useful context manager in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/tests/utils.py#L9.  Especially if it will be used by multiple test classes.